truffle/src/main/java/org/jruby/truffle/nodes/core/BigDecimalNodes.java

```diff
@@ -713,9 +729,124 @@ public Object div(RubyBasicObject a, RubyBignum b) {
                 "isRubyBigDecimal(b)",
                 "isNormal(b)"})
         public Object div(RubyBasicObject a, RubyBasicObject b) {
-            return divBigDecimal(a, getBigDecimalValue(b));
+            return div(a, getBigDecimalValue(b));
+        }
+
+        @Specialization(guards = {
+                "isNormal(a)",
+                "isRubyBigDecimal(b)",
+                "!isNormal(b)"})
+        public Object divNormalSpecial(RubyBasicObject a, RubyBasicObject b) {
+            switch (getBigDecimalType(b)) {
+                case NAN:
+                    return b;
+                case NEGATIVE_ZERO:
+                    switch (getBigDecimalValue(a).signum()) {
+                        case 1:
+                            return createRubyBigDecimal(Type.NEGATIVE_INFINITY);
+                        case 0:
+                            return createRubyBigDecimal(Type.NAN);
+                        case -1:
+                            return createRubyBigDecimal(Type.POSITIVE_INFINITY);
+                    }
+                case POSITIVE_INFINITY:
+                    switch (getBigDecimalValue(a).signum()) {
+                        case 1:
+                        case 0:
+                            return createRubyBigDecimal(BigDecimal.ZERO);
+                        case -1:
+                            return createRubyBigDecimal(Type.NEGATIVE_ZERO);
+                    }
+                case NEGATIVE_INFINITY:
+                    switch (getBigDecimalValue(b).signum()) {
+                        case 1:
+                            return createRubyBigDecimal(Type.NEGATIVE_ZERO);
+                        case 0:
+                        case -1:
+                            return createRubyBigDecimal(BigDecimal.ZERO);
+                    }
+            }
```

Comment:
yeah, I was thinking about that. I not happy with the current state. When I go back to add the casting node I should be able to improve this without specialization explosion.